auth/src/main/java/com/firebase/ui/auth/provider/FacebookProvider.java

```diff
@@ -184,6 +163,44 @@ public void onError(FacebookException error) {
         Bundle extra = new Bundle();
         extra.putString(ERROR, "error");
         extra.putString(ERROR_MSG, error.getMessage());
-        mCallbackObject.onFailure(extra);
+        onFailure(extra);
+    }
+
+    private IdpResponse createIdpResponse(String email, LoginResult loginResult) {
+        return new IdpResponse(
+                FacebookAuthProvider.PROVIDER_ID,
+                email,
+                loginResult.getAccessToken().getToken());
+    }
+
+    private void onSuccess(String email, LoginResult loginResult) {
+        gcCallbackManager();
+        mCallbackObject.onSuccess(createIdpResponse(email, loginResult));
+    }
+
+    private void onFailure(Bundle bundle) {
+        gcCallbackManager();
+        mCallbackObject.onFailure(bundle);
+    }
+
+    private void gcCallbackManager() {
+        // sCallbackManager must be static to prevent it from being destroyed if the activity
+        // containing FacebookProvider dies.
+        // In startLogin(Activity), LoginManager#registerCallback(CallbackManager, FacebookCallback)
+        // stores the FacebookCallback parameter--in this case a FacebookProvider instance--into
+        // a HashMap in the CallbackManager instance, sCallbackManager.
+        // Because FacebookProvider which contains an instance of an activity, mCallbackObject,
+        // is contained in sCallbackManager, that activity will not be garbage collected.
+        // Thus, we have leaked an Activity.
```

Comment:
Thanks for the super-detailed investigation.  I am fine with the long comment, it makes people more scared to undo your fix.